### PR TITLE
Add PageRank

### PR DESCRIPTION
**Pull request by IvanIsCoding**
Related to #315

Adds an implementation of the PageRank algorithm using sparse matrices. It uses the `sprs` crate combined with `ndarray` to implement a Power Method approach of finding the PageRank.

Also, we test this implementation against NetworkX's implementation of the PageRank. We accept all the arguments that NetworkX accepts: tolerance, max_iter, personalization, dangling, etc.

n.b: it's ready for review

### CODE REVIEW

**Comment by IvanIsCoding** [comment]
The current code works but does not handle dangling nodes. I will try to tweak that so we have a similar behaviour to NetworkX

**Review by mtreinish — COMMENTED** [review]
This is great, thanks for writing this! I really like using sprs for this, it also avoids us having to figure out how to link against and leverage blas/lapack to compute the eigenvectors of the google matrix from an ndarray. I need to refresh my memory on the algorithm before I do a detailed review on the algorithm code. I just has some quick high level comments from a quick scan of the code.

File: src/link_analysis.rs [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    py: Python,
+    graph: &PyDiGraph,
+    alpha: f64,
+    weight_fn: Option<PyObject>,
+    nstart: Option<HashMap<usize, f64>>,
+    personalization: Option<HashMap<usize, f64>>,
+    tol: f64,
+    max_iter: usize,
+    dangling: Option<HashMap<usize, f64>>,
+) -> PyResult<CentralityMapping> {
```
[the remark this answers is not in the document]
Comment (IvanIsCoding) [reply src/link_analysis.rs:97]: I can migrate PageRank, HITS to rustworkx-core once we add ndarray to their dependencies. It’s similar to why we didn’t migrate Floyd Warshall Numpy before

**Comment by IvanIsCoding** [comment]
> This is great, thanks for writing this! I really like using sprs for this, it also avoids us having to figure out how to link against and leverage blas/lapack to compute the eigenvectors of the google matrix from an ndarray. I need to refresh my memory on the algorithm before I do a detailed review on the algorithm code. I just has some quick high level comments from a quick scan of the code.

The algorithm is approximating the eigencector of the transition matrix. You might want to check NetworkX’s Python code directly because they have some quirks on how they handle dangling nodes etc.

File: src/link_analysis.rs [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+/// :rtype: CentralityMapping
+#[pyfunction(
+    alpha = "0.85",
+    weight_fn = "None",
+    nstart = "None",
+    personalization = "None",
+    tol = "1.0e-6",
+    max_iter = "100",
+    dangling = "None"
+)]
```
[the remark this answers is not in the document]
Comment (IvanIsCoding) [reply src/link_analysis.rs]: Done

File: .github/workflows/wheels.yml [hunk 3]
```diff
@@ -72,7 +72,7 @@ jobs:
           CIBW_MANYLINUX_I686_IMAGE: quay.io/pypa/manylinux2014_i686:latest
           CIBW_SKIP: cp36-* pp* *win32 *musl*
           CIBW_BEFORE_BUILD: pip install -U setuptools-rust
-          CIBW_TEST_REQUIRES: networkx testtools fixtures
+          CIBW_TEST_REQUIRES: networkx scipy testtools fixtures
```
[the remark this answers is not in the document]
Comment (IvanIsCoding) [reply .github/workflows/wheels.yml:75]: Done

**Review by mtreinish — COMMENTED** [review]
Sorry for the delay in review, this looks excellent to me. The code LGTM and nothing real stands out to me as being incorrect. Just a couple small inline suggestions and questions but other than I think this is ready to merge.

File: releasenotes/notes/add-pagerank-bef0de7d46026071.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            graph[node] = pagerank_score
+            colors.append(pagerank_score)
+        mpl_draw(
+            graph,
+            with_labels=True,
+            node_color=colors,
+            node_size=650,
+            labels=lambda x: "{0:.2f}".format(x)
+        )
+
```
[the remark this answers is not in the document]
Comment (IvanIsCoding) [reply releasenotes/notes/add-pagerank-bef0de7d46026071.yaml:29]: In theory `FailedToConverge` was introduced in #621, so it has been released for a while! I actually copied some code from your eigenvector centrality at the start

File: src/link_analysis.rs [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+///     not specified a default value of 100 is used.
+/// :param dict dangling: An optional dictionary for the outedges to be assigned to any "dangling" nodes,
+///     i.e., nodes without any outedges.  The dict key is the node the outedge points to and the dict
+///     value is the weight of that outedge. By default, dangling nodes are given outedges according to
+///     the personalization vector (uniform if not specified). This must be selected to result in an irreducible
+///     transition matrix. It may be common to have the dangling dict to be the same as the personalization dict.
+///
+/// :returns: a read-only dict-like object whose keys are the node indices and values are the
+///      PageRank score for that node.
+/// :rtype: CentralityMapping
```
[the remark this answers is not in the document]
Comment (IvanIsCoding) [reply src/link_analysis.rs:74]: I think I will add the notes for this, HITS, eigenvector centrality and all the centralities that cannot converge in a separate PR

File: Cargo.lock [hunk 6]
```diff
@@ -234,6 +284,16 @@ dependencies = [
  "num-traits",
 ]
 
+[[package]]
+name = "num-complex"
+version = "0.2.4"
```
[the remark this answers is not in the document]
Comment (IvanIsCoding) [reply Cargo.lock:289]: I think it they bumped the dependency recently. So maybe in the future? https://github.com/sparsemat/sprs/blob/master/sprs/Cargo.toml#L26

File: releasenotes/notes/add-pagerank-bef0de7d46026071.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (mtreinish) [reply releasenotes/notes/add-pagerank-bef0de7d46026071.yaml:29]: Lol, well apparently I completely forgot about that. Then yeah there is no need for a release note here.

**Comment by bionicles** [comment]
somehow i'm getting uniform ranks for the UMLS ontology (MRREL) with 45,891,578 relations, idk, I am using same exact arguments and graph as networkx but every rank is the same (1.0895245310588362e-08) when I use rustworkx 

```
    import rustworkx as rx
    
    env.debug("relations_df.shape:\n", relations_df.shape)
    graph = rx.PyDiGraph()

    env.debug("building a graph:")

    # Convert DataFrame columns to tuples for faster iteration
    env.debug("convert cui1 to a tuple")
    cui1_tuple = tuple(relations_df["cui1"])
    env.debug("converted cui1 to a tuple")

    env.debug("convert cui2 to a tuple")
    cui2_tuple = tuple(relations_df["cui2"])
    env.debug("converted cui2 to a tuple")

    # skipping for now bc unused for performance
    # env.debug("convert relations to a tuple")
    # relation_tuple = tuple(relations_df["relation"])
    # env.debug("converted relations to a tuple")

    total = len(cui1_tuple)
    for i in track(
        range(total),
        description="Constructing UMLS Relations Graph...",
    ):
        # graph.add_edges_from(((cui1, cui2), (cui2, cui1)))
        node_1_index = graph.add_node(cui1_tuple[i])
        node_2_index = graph.add_node(cui2_tuple[i])
        graph.add_edge(node_1_index, node_2_index, 1)
    return graph
    
def pagerank_relations_rx(
    graph: Optional[rx.PyDiGraph] = None,
    alpha: float = 0.85,
    personalization: Optional[dict] = None,
    max_iter: Optional[int] = 100,
    tol: Optional[float] = 1.0e-9,
    nstart: Optional[dict] = None,
    weight_fn: str = None,
    dangling: Optional[dict] = None,
) -> pd.DataFrame:
    "create an initial, unpersonalized ranking for the whole database"
    env.debug("page_rank_relations")
    if graph is None:
        graph = construct_relation_graph(
            env=env,
        )
    env.debug("graph:", graph)
    try:
        env.debug("Ranking with PageRank")
        rank_dict = rx.pagerank(
            graph,
            # alpha=alpha,
            # personalization=personalization,
            # max_iter=max_iter,
            # tol=tol,
            # nstart=nstart,
            # weight_fn=weight_fn,
            # dangling=dangling,
        )
        rank_dict_items = rank_dict.items()
        for _ in range(3):
            print(next(rank_dict_items))
        env.debug("PageRank Complete!")
    except Exception as e:
        message = f"page_rank_relations: nx.pagerank Exception e: {type(e)}"
        env.error(message)
        raise Exception(message)
    # separate exception for sorting
    try:
        env.debug("Sorting...")
        records = tuple(rank_dict.items())
        for record_number in range(3):
            print(records[record_number])
        for record_number in range(3):
            print(records[-record_number])
        # ... irrelevant
    ```
 ```
[2024-01-17 19:02:55.490423-05:00 environment.py::361]
> WARN (read_relations): Loading relations, this may take a moment...

[2024-01-17 19:03:02.660546-05:00 environment.py::371]
> WARN (read_relations): Loaded 45,891,578 relations in 7.2 seconds!

[2024-01-17 19:03:02.661256-05:00 rank.py::171]
> DEBUG (construct_relation_graph_rx): construct_relation_graph

[2024-01-17 19:03:02.661709-05:00 rank.py::175]
> DEBUG (construct_relation_graph_rx): relations_df.shape:

(45891578, 3)

[2024-01-17 19:03:02.662325-05:00 rank.py::178]
> DEBUG (construct_relation_graph_rx): building a graph:

[2024-01-17 19:03:02.662678-05:00 rank.py::181]
> DEBUG (construct_relation_graph_rx): convert cui1 to a tuple

[2024-01-17 19:03:08.934134-05:00 rank.py::183]
> DEBUG (construct_relation_graph_rx): converted cui1 to a tuple

[2024-01-17 19:03:08.934678-05:00 rank.py::185]
> DEBUG (construct_relation_graph_rx): convert cui2 to a tuple

[2024-01-17 19:03:15.181439-05:00 rank.py::187]
> DEBUG (construct_relation_graph_rx): converted cui2 to a tuple
Constructing UMLS Relations Graph... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:12

[2024-01-17 19:03:28.079813-05:00 rank.py::119]
> DEBUG (pagerank_relations_rx): page_rank_relations

[2024-01-17 19:03:28.080357-05:00 rank.py::124]
> DEBUG (pagerank_relations_rx): graph: <rustworkx.PyDiGraph object at 0x7f4dfa2e4e80>

[2024-01-17 19:03:28.080856-05:00 rank.py::126]
> DEBUG (pagerank_relations_rx): Ranking with PageRank
(0, 1.0895245310588362e-08)
(1, 1.0895245310588362e-08)
(2, 1.0895245310588362e-08)

[2024-01-17 19:03:41.795034-05:00 rank.py::140]
> DEBUG (pagerank_relations_rx): PageRank Complete!

[2024-01-17 19:03:41.795569-05:00 rank.py::147]
> DEBUG (pagerank_relations_rx): Sorting...
(0, 1.0895245310588362e-08)
(1, 1.0895245310588362e-08)
(2, 1.0895245310588362e-08)
(0, 1.0895245310588362e-08)
(91783155, 1.0895245310588362e-08)
(91783154, 1.0895245310588362e-08)
```

**Comment by IvanIsCoding** [comment]
> gonna put a pin in it and double check everything and maybe make a new issue about it, but somehow i'm getting totally uniform results from this where networkx is (a lot) slower but produces some results that make more sense

Please open a new issue. If you could give the graph that triggers the difference that would be great as well.

We use the Power Method and NetworkX uses SVD decomposition to approximate the eigenvalue so there might be some discrepancies.

**Pull request merged**